### ISSUE TITLE
Training patch experts using cen model on NIR Images

### ISSUE THREAD

**Issue by Nourhan27** [TadasBaltrusaitis/OpenFace#745]
Hello Tadas,

I used ce-clm code to train patch experts on NIR dataset. The training finished successfully but i got very bad accuracy although i completed the training on the old rgb (In the wild) model weights.

Visually the output using the rgb model on NIR images is much better than the new trained model.

Do you have any justification for this ?

Thanks,

**Comment by TadasBaltrusaitis** [comment]
It's hard to tell, there are a lot of potential issues: incorrect data formating/processing; small data size; not enough time spent optimizing - the model failed to converge etc.

Some questions that might help you debug:
1. Did you start with RGB images and then tune on NIR data?
2. Did the training loss on NIR data get better? 
3. Was the issue with generalization, the model worked well on training/validation but failed on test set
4. Did you visualize the heat maps on the test data to see what the NIR patch experts are doing compared to the RGB ones?

**Comment by Nourhan27** [comment]
Hello Tadas,

First Thanks for your reply, highly appreciated.
1- I used 1600 NIR training sample, i used prepare_data_wild_all to prepare data (The data is classified to frontal and profile1 view) ,  then i used Generate_Patches_Wild.m to generate patches for each scale.

2- I used cen_patches_0.25_of.mat , cen_patches_0.35_of.mat , cen_patches_0.50_of.mat , cen_patches_1.00_of.mat to extract the weights of these models after that , i used these weights and initialized the network before training with the extracted weights.  I just loaded the weights but i miss the optimizer weights because it does not exist in *.mat files.

3- I started training for each view (frontal , profile1) , scale (0.25,0.35,0.5,1)  , landmark for 50 epochs each, using 512 minibatch size.

4- After fininshing the training i used create_cen_experts_wild.m to generate the models in mat then dat format.

** But i noticed that starting first epoch till epoch number 50 the rmse is almost fixed around 0.08... and corr is very small around 0.0001 and keeps decreasing to 10^(-5), and loss is almost the same aroung 0.007,  so i think that from the beginning of the training there is something wrong with rmse and corr values. But i continued it maybe it changes, but it did not.

** The model is bad for training and test.

**Did you visualize the heat maps on the test data to see what the NIR patch experts are doing compared to the RGB ones?**

How could i do this ?? 

Thanks ,

**Comment by TadasBaltrusaitis** [comment]
Hi,

Some feedback:
1. One thing that I find useful after running prepare_data_wild_all and Generate_Patches_Wild is to actually visualize the patches of certain landmarks to see that the extraction scripts actually extracted what was intended (e.g. patches with 1 annotation are actually landmarks of interest and patches with 0 annotation are not).
2. Can you recreate the training of CEN with RGB patches? That is if you take the same steps but for RGB data can you train the model successfully (it is always easier to go from working model to working model). For experimentation I would pick a landmark that is "easy", for example corner of the eye and see if you can get good correlation and RMSE values with RGB data and then move onto NIR data.
3. In case it turns out that your NIR images are not enough (e.g. they're not diverse enough etc.), you can always try mixing RGB + NIR data
4 For visualizing the heat maps, there's several ways. You can extract the matlab models and put them in the matlab framework. E.g. start with `\demo\face_image_demo.m` and see where the patch experts are actually applied to the image, specifically line 147 of `Fitting_from_bb.m` that applies the CEN model on patches, and you can visualize the response map to see what it should look like (start with seeing what it looks like for RGB data before trying with NIR).
5. Do the initially loaded weights from RGB model produce good RMSE and correlation metrics? Are they good at the beginning and then get worse? If they are bad at the beginning then something went wrong with data loading or data generation; if they start good and get worse when trained on your data then something is wrong with optimization parameters.
